source/common/http/http1/codec_impl.cc

```diff
@@ -158,16 +158,17 @@ void StreamEncoderImpl::encodeHeadersBase(const RequestOrResponseHeaderMap& head
       // response to a HEAD request.
       // For 204s and 1xx where content length is disallowed, don't append the content length but
       // also don't chunk encode.
-      if (is_content_length_allowed_) {
+      if (!is_1xx_or_204_) {
         encodeFormattedHeader(Headers::get().ContentLength.get(), "0");
       }
       chunk_encoding_ = false;
     } else if (connection_.protocol() == Protocol::Http10) {
       chunk_encoding_ = false;
     } else {
       // For responses to connect requests, do not send the chunked encoding header:
-      // https://tools.ietf.org/html/rfc7231#section-4.3.6
-      if (!is_response_to_connect_request_) {
+      // https://tools.ietf.org/html/rfc7231#section-4.3.6. The same is true for 1xx and 204
+      // responses: https://tools.ietf.org/html/rfc7230#section-3.3.1
+      if (!is_response_to_connect_request_ && !is_1xx_or_204_) {
```

Comment:
The one thing that makes me a little twitchy, is that we're setting chunk_encoding_ for 1xx despite the fact we don't want to.

Arguably, this is fine, because when the informational headers arrive, we'll run through this function again and set chunk_encoding_ to its final value.  

Optionally, it might be cleaner to just avoid the else block for all of 1xx_or_204 since for 1xx it shouldn't change the chunk_encoding_state and for 204 it should be moot (and we could even assert it's the end of the stream).  

I think given that chunk_encoding_ defaults true I'd like an assert that this is the end of the stream either way, since chunk_encoding_ defaults to true and we want to make sure if there's ever some weird corner case of frame-by-connection-close we properly don't send a 0 byte chunked body.

Reply:
Done. I ended up splitting out 101/204 into separate flags to avoid re-parsing the status header.